### PR TITLE
Chore/strapi database lifecycle

### PR DESCRIPTION
**Pull request by alexandrebodin**
This PR moves the lifecycle management to the strapi-database layer.

### Advantages
This fixes issues with the order of lifecycles events and improves usability. 

### Tradeoffs
Doing this in our own layer means a user won't be able to use lifecycles when using an ORM model manually:

This will not trigger the lifecycle hooks
```js
strapi.query('restaurant').model.forge().save()
```

But this will trigger the lifecycle hooks
```js
strapi.query('restaurant').create()
```

### Events

We renamed the lifecycle events to be consistent with the strapi `queries` layer.

- `beforeCreate`
- `afterCreate`
- `beforeUpdate`
- `afterUpdate`
- `beforeDelete`
- `afterDelete`
- `beforeFind`
- `afterFind`
- `beforeFindOne`
- `afterFindOne`
- `beforeCount`
- `afterCount`
- `beforeSearch`
- `afterSearch`
- `beforeCountSearch`
- `afterCountSearch`

### Notes

- All `before` events will get the params coming from the `strapi.query(model)[method](...args)` call
- All `after` events will get the result from the query, and the previous parameters.

### Todo

- [x] Update model templates
- [x] Add documentation on each hook arguments
- [x] Might need to refacto the query arguments so it is easier to work with.

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
Well that was fast

**Comment by aksdevac** [comment]
Hi, is there any way to do it without mentioned trade offs ? or is there any way to use both ORM hooks & this?

**Comment by derrickmehaffy** [comment]
> Hi, is there any way to do it without mentioned trade offs ? or is there any way to use both ORM hooks & this?

No, the lifecycle callbacks have been broken due to those ORM calls. Both of the core ORMs Strapi uses (Mongoose and Bookshelf) have more or less dumped, removed, or broke their own lifecycles. (Mongoose comes to mind removing delete). 

Hence why this PR is moving it to Strapi logic

**Comment by alexandrebodin** [comment]
@aksdevac In the future we will slow but surely remove the need for the underlying ORM so the query layer is a lot more flexible and powerful.

If we were to combine both it would make no sense for the users which one to choose and might create issue as both would be called at the same time. Moreover the ORM specific lifecycles don't have the same API's depending on the connector and this is also an issue. 

Can you give a usecase where you would absoltely need the ORM lifecycles instead of adding the code manually as it is already custom code when you use it ?

**Comment by derrickmehaffy** [comment]
> @aksdevac In the future we will slow but surely remove the need for the underlying ORM so the query layer is a lot more flexible and powerful.
> 
> If we were to combine both it would make no sense for the users which one to choose and might create issue as both would be called at the same time. Moreover the ORM specific lifecycles don't have the same API's depending on the connector and this is also an issue.
> 
> Can you give a usecase where you would absolutely need the ORM lifecycles instead of adding the code manually as it is already custom code when you use it ?

I can think of a few myself but I would just be writing custom before/after stuff in the controllers anyway. Mainly geared towards complex queries such as joins in SQL.

**Comment by aksdevac** [comment]
> @aksdevac In the future we will slow but surely remove the need for the underlying ORM so the query layer is a lot more flexible and powerful.
> 
> If we were to combine both it would make no sense for the users which one to choose and might create issue as both would be called at the same time. Moreover the ORM specific lifecycles don't have the same API's depending on the connector and this is also an issue.
> 
> Can you give a usecase where you would absolutely need the ORM lifecycles instead of adding the code manually as it is already custom code when you use it ?

ok, I don't think both will be called at same time in all cases, as you said if we use underlying model directly for create/update starpi's hook won't be called. but it will create confusion, but if it can be supported using optinal thing it would be good. as I think this issue is not present if we use mongoose directly without starpi.

File: packages/strapi-connector-bookshelf/lib/mount-models.js [hunk 1]
```diff
@@ -601,37 +570,12 @@ module.exports = ({ models, target }, ctx) => {
           });
         }
 
-        function formatOutput(instance) {
+        this.on('saved fetched fetched:collection', instance => {
```
Comment (stafyniaksacha) [inline packages/strapi-connector-bookshelf/lib/mount-models.js:573]: Hello @alexandrebodin, what is this  `'saved fetched fetched:collection'` mean ? (it's a curriosity)

Comment (alexandrebodin) [reply packages/strapi-connector-bookshelf/lib/mount-models.js:573]: This is bookshelf syntax to listen on multiple events. Here we transform the output of find / findOne and after saving an entity

File: docs/3.0.0-beta.x/concepts/models.md [hunk 2]
```diff
@@ -979,43 +1102,43 @@ module.exports = {
...
-
-    // Set the password.
-    model.set('password', passwordHashed);
+  lifecycles: {
+    beforeCreate(data) {
+      data = {
+        ...data,
+        name: 'Some fixed name',
+      };
+    },
```
Comment (petersg83) [inline docs/3.0.0-beta.x/concepts/models.md:1139]: Good idea to warn about this

File: docs/3.0.0-beta.x/concepts/models.md [hunk 3]
```diff
@@ -1142,3 +1142,28 @@ module.exports = {
...
 
 :::
+
+### Custom use
+
+When you are building custom ORM specific queries the lifecycles will not be triggered. You can however call a lifecycle function directly if you wish.
+
+**Bookshelf example**
+
+**Path -** `someService.s`
```
Comment (petersg83) [inline docs/3.0.0-beta.x/concepts/models.md:1152]: .js

**Comment by roelbeerens** [comment]
This is awesome!

**Pull request merged**